src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/TempDataApplicationModelProvider.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.ApplicationModels;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public class TempDataApplicationModelProvider : IApplicationModelProvider
+    {
+        /// <inheritdoc />
+        public int Order { get { return -1000 + 10; } }
+
+        /// <inheritdoc />
+        public void OnProvidersExecuted(ApplicationModelProviderContext context)
+        {
+        }
+
+        /// <inheritdoc />
+        public void OnProvidersExecuting(ApplicationModelProviderContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            foreach (var controllerModel in context.Result.Controllers)
+            {
+                var properties = controllerModel.ControllerType.GetProperties();
```

Comment:
We should use https://github.com/aspnet/Common/blob/dev/shared/Microsoft.Extensions.PropertyHelper.Sources/PropertyHelper.cs#L166 here